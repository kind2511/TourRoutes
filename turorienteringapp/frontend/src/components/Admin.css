.admin-container {
    display: flex;
    align-items: stretch;
    padding: 20px;
    overflow: hidden; /* To handle overflow of content */
}

.admin-column {
    flex: 1;
    margin: 0 20px;
    padding: 10px;
    background-color: #f5f5f5;
    border-radius: 5px;
    box-shadow: 0 2px 4px rgba(39, 234, 9, 0.911);
    overflow-y: auto; /* Allows scrolling within each column */
}

.admin-column h2 {
    margin-bottom: 10px;
}

.admin-column ul {
    list-style-type: none;
    padding: 0; /* Space*/
}

.admin-column ul li {
    border-bottom: 1px solid  #4CAF50;
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 13px 0; /* space between elements vertically */
    }

.admin-column ul li:last-child {
    border-bottom: none;
}

/* Style for the divider between users and routes*/
.admin-column-divider {
    background-color:  #4CAF50;;
    width: 2px;
    margin: 0 10px;
}


/* Style for the address of coloms */
.admin-column h2 {
    font-size: 40px;
    font-style: italic;
    color:rgb(255, 0, 149);
}

.admin-routes-delete-button {
    color:white;
    background-color:red;
    width:100px;
    margin-right:200px;
    padding: 1px 0px 0px; /*This will make the button smaller*/
    }