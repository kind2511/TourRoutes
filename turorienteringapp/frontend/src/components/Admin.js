import React, { useEffect, useState } from 'react';
import './Admin.css';
import AdminDeleteUser from './AdminDeleteUser';
import AdminDeleteRoute from './AdminDeleteRoute';


const Admin = () => {
    const [users, setUsers] = useState([]);
    const [routes, setRoutes] = useState([]);

    //------------------------------------------->

    // Fetch Users from the backend  and set them in colom lista 
    const fetchUsers = async () => {
        try {
            // Retrieve the token from local storage
            const token = localStorage.getItem('token');

            // If there's no token, throw an error or handle as needed
            if (!token) {
                throw new Error('No token found');
            }

            const response = await fetch('http://localhost:8000/api/v1/users', {
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            });
            const data = await response.json();

            if (response.ok) {
                setUsers(data.data.users);
            } else {
                // If the status code is not ok, log the status and the message
                console.error(`Failed to fetch users: ${data.message}`);
            }
        } catch (error) {
            console.error('Error fetching users:', error);
        }
    };

    //------------------------------------------------->

    // Fetch Routes from the backend
    const fetchRoutes = async () => {
        try {
            const token = localStorage.getItem('token');
            if (!token) {
                throw new Error('Authentication token not found');
            }
            const response = await fetch('http://localhost:8000/api/v1/tourRoutes', {
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            });
            const data = await response.json();
            if (response.ok) {
                setRoutes(data.data.tourRoutes);
            }
        } catch (error) {
            console.error('Error fetching routes:', error);
        }
    };

    // Call both functions once the when the component first mounted 
    useEffect(() => {
        fetchUsers();
        fetchRoutes();
    }, []);

    //------------------------------------------------------------>

    /* 
     * Function to delete route by the Admin 
     */

    const handleDeleteRoute = async (routeId) => {
        try {
            const token = localStorage.getItem('token');
            if (!token) {
                console.error('No authentication token found');
                return;
            }

            const response = await fetch(`http://localhost:8000/api/v1/tourRoutes/${routeId}`, {
                method: 'DELETE',
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            });

            const data = await response.json();
            if (response.ok) {
                console.log("Route deleted successfully:", data);


                setRoutes(routes.filter(route => route._id !== routeId)); // UI filter to remove the route directly from screen
            } else {
                console.error('Failed to delete route:', data.message);
            }
        } catch (error) {
            console.error('Error deleting route:', error);
        }
    };

    //------------------------------------------------------------>

    /* 
     * Function to delete user by the Admin 
     */

    const handleDeleteUser = async (userId) => {
        try {
            const token = localStorage.getItem('token');
            if (!token) {
                console.error('No authentication token found');
                return;
            }

            const response = await fetch(`http://localhost:8000/api/v1/users/${userId}`, {
                method: 'DELETE',
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            });

            const data = await response.json();
            if (response.ok) {
                console.log("User deleted successfully:", data);

                setUsers(users.filter(user => user._id !== userId)); // UI filter to remove the user directly from screen
            } else {
                console.error('Failed to delete user:', data.message);
            }
        } catch (error) {
            console.error('Error deleting user:', error);
        }
    };

    //------------------------------------------------------------>



    return (
        <div className="admin-container">
            <div className="admin-column">
                <h2>Users:</h2>
                <ul>
                    {users.map((user, index) => (
                        <li key={user._id}>
                            {index + 1}. {user.firstName} {user.lastName}
                            <div className="admin-button-container">{/*Added div for each user list item*/}
                                <AdminDeleteUser userId={user._id} onDelete={handleDeleteUser} />

                                <button className="admin-users-promote-button">Promote</button>
                            </div>
                        </li> // Render users as a list
                    ))}
                </ul>
            </div>

            <div className="admin-column-divider"></div>

            <div className="admin-column">
                <h2>Routes:</h2>
                <ul>
                    {routes.map((route, index) => (
                        <li key={route._id}>
                            {index + 1}. {route.name}, ID: {route._id}
                            <AdminDeleteRoute routeId={route._id} onDelete={handleDeleteRoute} />
                        </li> // Render routes as a list
                    ))}
                </ul>
            </div>
        </div>
    );
};

export default Admin;