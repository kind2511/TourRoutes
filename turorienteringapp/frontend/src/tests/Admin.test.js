import React from 'react';
import { render } from '@testing-library/react';
import {BrowserRouter} from 'react-router-dom';
import Admin from '../components/Admin';

test('Admin component renders without crashing', () => {
  render(<Admin />);
});

//---------------------------------------->

// Snapshot test for the Admin component
test('Admin component snapshot', () => {
  const { asFragment } = render(
    <BrowserRouter>
      <Admin />
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});