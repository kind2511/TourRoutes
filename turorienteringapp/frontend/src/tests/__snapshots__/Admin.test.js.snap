// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin component snapshot 1`] = `
<DocumentFragment>
  <div
    class="admin-container"
  >
    <div
      class="admin-column"
    >
      <h2>
        Users:
      </h2>
      <ul />
    </div>
    <div
      class="admin-column-divider"
    />
    <div
      class="admin-column"
    >
      <h2>
        Routes:
      </h2>
      <ul />
    </div>
  </div>
</DocumentFragment>
`;
